packages/node-utils/src/http.ts

```diff
@@ -0,0 +1,324 @@
+                    this.logger.info('trying to close server which was not running');
+                }
+                this.logger.info('Server stopped');
+                this.emitter.emit('server/closed');
+                resolve();
+            });
+            Object.values(this.sockets).forEach(socket => {
+                socket.destroy();
+            });
+        });
+    }
+
+    public post(pathname: string, handler: Handler[]) {
+        this.routes.push({
+            method: 'POST',
+            pathname,
+            handler,
+        });
+    }
+
+    public get(pathname: string, handler: Handler[]) {
+        this.routes.push({
+            method: 'GET',
+            pathname,
+            handler,
+        });
+    }
+
+    /**
+     * Register common handlers that are run for all requests before route handlers
+     */
+    public use(handler: Handler[]) {
+        this.routes.push({
+            method: '*',
+            pathname: '*',
+            handler,
+        });
+    }
+
+    // PUT, DELETE etc are not used anywhere in our codebase, so no need to implement them now
+
+    /**
+     * Entry point for handling requests
+     */
+    private onRequest = (request: http.IncomingMessage, response: http.ServerResponse) => {
+        // mostly ts stuff. request should always have url defined.
+        if (!request.url) {
+            this.logger.warn('Unexpected incoming message (no url)');
+            this.emitter.emit('server/error', 'Unexpected incoming message');
+            return;
+        }
+
+        request.on('aborted', () => {
+            this.logger.info(`Request ${request.url} aborted`);
+        });
+
+        const { pathname } = url.parse(request.url, true);
+
+        this.logger.info(`Handling request for ${pathname}`);
+
+        const route = this.routes.find(r => r.pathname === pathname && r.method === request.method);
+        if (!route) {
+            this.emitter.emit('server/error', `Route not found for ${pathname}`);
+            this.logger.warn(`Route not found for ${pathname}`);
+            return;
+        }
+
+        if (!route.handler.length) {
+            this.emitter.emit('server/error', `No handlers registered for route ${pathname}`);
+            this.logger.warn(`No handlers registered for route ${pathname}`);
+            return;
+        }
+
+        const handlers = [
+            ...this.routes
+                .filter(r => r.method === '*' && r.pathname === '*')
+                .flatMap(r => r.handler),
+            ...route.handler,
+        ];
+
+        let i = 0;
+        const run = (handler: Handler) => {
+            handler(
+                request as Request,
+                response,
+                () => {
+                    i++;
+
+                    if (handlers[i]) {
+                        run(handlers[i]);
+                    }
+                },
+                { logger: this.logger },
+            );
+        };
+
+        run(route.handler[0]);
```

Comment:
I think that the problem was solely in calling `run(route.handler[0])` instead of `run(handlers[0])`. But I've also tried to polish it a bit.

```suggestion
        const run = ([handler, ...rest]: Handler[]) => {
            handler?.(request as Request, response, () => run(rest), { logger: this.logger });
        };

        run(handlers);
```